File: 0904-fruit-into-baskets/0904-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        left, n = 0, len(fruits)
        ans, seen = 0, defaultdict(int)

        for right in range(n):
            while fruits[right] not in seen and len(seen) >= 2:
                seen[fruits[left]] -= 1
                if seen[fruits[left]] == 0:
                    del seen[fruits[left]]
                left += 1
            
            seen[fruits[right]] += 1
            ans = max(ans, right - left + 1)

        return ans

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int left = 0, n = fruits.size(), ans = 0;
        unordered_map<int, int> seen;
        for (int right = 0; right < n; ++right){
            while (seen.count(fruits[right]) == 0 && seen.size() >= 2){
                --seen[fruits[left]];
                if (seen[fruits[left]] == 0) seen.erase(fruits[left]);
                ++left;
            }

            ++seen[fruits[right]];
            ans = max(ans, right - left + 1);
        }

        return ans;
    }
};